src/ProjectOrigin.WalletSystem.Server/Repositories/CertificateRepository.cs

```diff
@@ -423,50 +423,41 @@ public async Task SetWalletSliceState(Guid sliceId, WalletSliceState state)
-            throw new InvalidOperationException($"Owner has no available slices to reserve");
+        var slices = await GetOwnersAvailableSlices(registryName, certificateId, owner);
+        var willBeAvailable = await GetRegisteringAndAvailableQuantity(registryName, certificateId, owner);
 
-        if (availableSlices.Sum(slice => slice.Quantity) < reserveQuantity)
+        if (slices.Sum(x => x.Quantity) >= reserveQuantity)
         {
-            var registeringAvailableQuantity =
-                await GetRegisteringAndAvailableQuantity(registryName, certificateId, owner);
-            if (registeringAvailableQuantity >= reserveQuantity)
-                throw new TransientException($"Owner has enough quantity, but it is not yet available to reserve");
-            else
-                throw new InvalidOperationException($"Owner has less to reserve than available");
-        }
+            var sumSlicesTaken = 0L;
+            var takenSlices = slices.Where(s => s.State == WalletSliceState.Available)
+                .OrderBy(slice => slice.Quantity)
+                .TakeWhile(slice =>
+                {
+                    var needsMore = sumSlicesTaken < reserveQuantity;
+                    sumSlicesTaken += slice.Quantity;
+                    return needsMore;
+                })
+                .ToList();
 
-        var sumSlicesTaken = 0L;
-        var takenSlices = availableSlices
-            .OrderBy(slice => slice.Quantity)
-            .TakeWhile(slice =>
+            foreach (var slice in takenSlices)
             {
-                var needsMore = sumSlicesTaken < reserveQuantity;
-                sumSlicesTaken += slice.Quantity;
-                return needsMore;
-            })
-            .ToList();
+                await SetWalletSliceState(slice.Id, WalletSliceState.Reserved);
+            }
 
-        foreach (var slice in takenSlices)
+            return takenSlices;
+        }
+        else if (willBeAvailable < reserveQuantity)
         {
-            var rowsChanged = await _connection.ExecuteAsync(
-                @"UPDATE wallet_slices
-                SET state = @state
-                WHERE id = @sliceId
-                    AND state = @expected",
-                new
-                {
-                    sliceid = slice.Id,
-                    state = WalletSliceState.Reserved,
-                    expected = WalletSliceState.Available
-                });
-
-            if (rowsChanged != 1)
-                throw new InvalidOperationException($"Slice with id {slice.Id} could not be found or was no longer available");
+            throw new InvalidOperationException($"Owner has less to reserve than available");
+        }
+        else if (willBeAvailable >= reserveQuantity)
+        {
+            throw new TransientException($"Owner has enough quantity, but it is not yet available to reserve");
+        }
+        else
+        {
+            throw new Exception("Unexpected error");
```

Comment:
Jeg ville nok foretrække vi lige var en tand mere specifik, ved godt det var mit eget

```suggestion
            throw new InvalidOperationException("This code path should never be reached.");
```

Reply:
Skal vi egentlig overhovedet sluge den? Vi vil jo gerne have hele stacktracet med vel

Reply:
Nok ikke fange den nej, men jeg synes casen giver mening, for hvis man skulle ændre i koden og køre forbi det hele, så har vi et problem.

Reply:
Hov, jeg tænkte det lå i en catch.. Jeg lader det være :)